File: app/src/main/java/com/thecoolguy/rumaan/fileio/data/MainActivityViewModel.java
package com.thecoolguy.rumaan.fileio.data;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.content.Context;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.util.Log;
import android.widget.Toast;
import com.thecoolguy.rumaan.fileio.R;
import com.thecoolguy.rumaan.fileio.data.models.LocalFile;
import com.thecoolguy.rumaan.fileio.data.repository.Repository;
import com.thecoolguy.rumaan.fileio.utils.Utils;
import org.jetbrains.annotations.NotNull;

/**
 * Abstraction for Repository with Lifecycle aware stuffs
 */

public class MainActivityViewModel extends AndroidViewModel {

  private static final String TAG = "MainActivityViewModel";

  /* Store the current chosen file instance */
  private LocalFile localFile;

  private MainActivityViewModel(@NonNull Application application) {
    super(application);
  }


  /**
   * Get the file from the database and save in the current view model state.
   */
  public void chooseFileFromUri(@NotNull final Context context, @NotNull final Uri fileUri) {
    this.localFile = Utils.getLocalFile(context, fileUri);
  }

  public void uploadFile(@NotNull final Context context) {
    if (localFile == null) {
      Log.e(TAG, "LocalFile object null! uploadFile() called before choosing a file!",
          new NullPointerException());
      Toast.makeText(context, context.getString(R.string.oops_some_error_occurred),
          Toast.LENGTH_SHORT).show();
    } else {
      Repository.initiateUpload(localFile);
    }
  }

}

File: app/src/main/java/com/thecoolguy/rumaan/fileio/data/MainActivityViewModel.kt
package com.thecoolguy.rumaan.fileio.data

import android.arch.lifecycle.ViewModel
import android.content.Context
import android.net.Uri
import android.util.Log
import com.thecoolguy.rumaan.fileio.data.models.LocalFile
import com.thecoolguy.rumaan.fileio.data.repository.Repository
import com.thecoolguy.rumaan.fileio.listeners.FileLoadListener
import com.thecoolguy.rumaan.fileio.listeners.FileUploadProgressListener
import com.thecoolguy.rumaan.fileio.utils.Utils
import kotlinx.coroutines.experimental.launch

/**
 * Abstraction for Repository with Lifecycle aware stuffs
 */

class MainActivityViewModel : ViewModel() {
    /* Store the current chosen file instance */
    private var localFile: LocalFile? = null

    /**
     * Get the file from the database and save in the current view model state.
     */
    fun chooseFileFromUri(context: Context, fileUri: Uri, fileLoadListener: FileLoadListener) {
        launch {
            localFile = Utils.getLocalFile(context, fileUri)
            localFile?.let {
                fileLoadListener.onFileLoad(it)
            }
        }
        Log.i(TAG, "Outside!")
    }

    /**
     * Upload the file to server and save the response into the database.
     * */
    fun uploadFile(listener: FileUploadProgressListener) {
        localFile?.let {
            // upload file
            Repository.getInstance().upload(it, listener)

            // TODO: update progress
        }
    }

    companion object {
        private val TAG = "MainActivityViewModel"
    }

}
